XML Parser 제작 / 데이터 DB 저장 로직 작성

Do-ho opened the pull request:
## 📑 제목

XML Parser 제작 / 데이터 DB 저장 로직 작성

![image](https://user-images.githubusercontent.com/33643752/113561759-176c9f00-9640-11eb-942c-60639e34f1ef.png)


## 💬 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] RestTemplate를 통한 과목 정보 받아오기
- [x] XML를 DOM으로 만들어 과목 ID와 정보 얻어오기
- [x] 데이터 DB 저장


## 🚧 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 부족한 코드지만 잘 부탁드려요...!
- 현재는 데이터를 삽입하는 것을 성공은 했습니다. 하지만 Test 코드를 작성하는데 어려움을 겪었어요 ㅠㅠ Getone으로 ID값을 통해 과목을 조회하고 싶었는데
```
Lecture lecture = lectureRepository.getOne(2021100863L);
```
- 해당 코드처럼 작성을 하면서 long 자료형에 맞춰서 getOne 조회를 하려 했습니다.
![image](https://user-images.githubusercontent.com/33643752/113562234-ba251d80-9640-11eb-8aac-0448823cfe80.png)
- 다음과 같은 오류가 발생하였고 어떤 식으로 조회를 해야할 지 감이 안잡혀서 고민 공유 남깁니다..!
- 현재 위의 현상은 Commit을 하지 않았습니다!
     - ```findbyid```를 통해 해결했습니다!

- 2번째로 배열에 있는 과목 데이터를 계속적으로 과목을 생성해 save를 시키는데 이래도 괜찮을까요?? ㅋㅋㅋ... ㅠㅠ
     - ```saveall```을 통해 해결했습니다!
- H2-Console 을 보기 위해 해당 설정을 했습니다. [이슈](https://pythonq.com/so/spring-boot/228997)

Do-ho committed:
feat: LectureGetter 구현

- 학사에 API 요청을 통한 과목 정보 얻는 로직 구현

Do-ho committed:
feat: XMLParser 구현

- xml 정보를 가진 string 타입의 변수를 DOM 트리로 만들기 위해 Parsing 하는 로직 작성

Do-ho committed:
test: XMLParserTest 작성

youngxpepp commented on src/main/java/com/github/hanpyo/util/LectureGetter.java:
```diff
@@ -0,0 +1,27 @@
+    private static String url = "https://kut90.koreatech.ac.kr/nexacroController.do";
+    private static String xmlString = "<Root xmlns=\"http://www.nexacroplatform.com/platform/dataset\"><Parameters><Parameter id=\"method\">getList_sp</Parameter><Parameter id=\"sqlid\">NK_SOT_MST.NP_SELECT_12</Parameter><Parameter id=\"programid\">SO0290S01</Parameter><Parameter id=\"locale\" /></Parameters><Dataset id=\"input1\"><ColumnInfo><Column id=\"YR\" type=\"string\" size=\"4000\" /><Col ...
+
+    public static RestTemplate restTpl = new RestTemplate();
```
RestTemplate은 Connection Pool을 사용해야 할 것 같아요.
https://nesoy.github.io/articles/2020-05/RestTemplate

또한 Bean으로 만드는건 어떨까요? 아래 링크를 참조하면 RestTemplate을 Bean으로 만들더라구요.
RestTemplate이 LectureGetter 뿐만 아니라 다른 곳에서 쓰일 수도 있고, RestTemplateBuilder를 이용해서 만들면 더 편한 장점도 있습니다. 
https://cheese10yun.github.io/spring-guide-api-call/

youngxpepp commented on src/main/java/com/github/hanpyo/util/LectureGetter.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+
+public abstract class LectureGetter {
```
Bean은 Bean에서만 쓸 수 있어요. RestTemplate이 Bean이 되면 LectureGetter도 Bean이 되겠죠? 
abstract 키워드말고 그냥 클래스로 선언하는건 어떨까요?

youngxpepp commented on src/main/java/com/github/hanpyo/util/LectureGetter.java:
```diff
@@ -0,0 +1,27 @@
+
+    public static RestTemplate restTpl = new RestTemplate();
+
+    public static String getLectureData() {
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.TEXT_XML);
+
+        HttpEntity<String> request = new HttpEntity<>(xmlString, headers);
+
+        final ResponseEntity<String> response = restTpl.postForEntity(url, request, String.class);
+
+        return response.getBody();
+    }
```
String으로 반환해주는 것보다 객체로 반환하는 것이 좋아보여요. XML을 객체로 변환하여 반환하면 해당 API가 정확히 어떤 데이터를 응답하는지 직관적으로 알 수 있을 것 같습니당. 아래 링크는 XML을 객체로 변환하는 글이에용ㅎㅎ

- https://blog.kingbbode.com/6
- https://zero-gravity.tistory.com/288

Do-ho replied:
추후에 바꿔보겠습니다!

youngxpepp commented on src/main/java/com/github/hanpyo/util/XMLParser.java:
```diff
@@ -0,0 +1,41 @@
+import javax.xml.parsers.DocumentBuilderFactory;
+import java.io.ByteArrayInputStream;
+
+public abstract class XMLParser {
+    public static Document parseXmlString(String src) throws Exception {
+        DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
+        builderFactory.setIgnoringElementContentWhitespace(true);
+
+        DocumentBuilder builder = builderFactory.newDocumentBuilder();
+        StringBuilder xmlStringBuilder = new StringBuilder();
+        xmlStringBuilder.append(src);
+
+        ByteArrayInputStream input = new ByteArrayInputStream(xmlStringBuilder.toString().getBytes("UTF-8"));
+        Document document = builder.parse(input);
+        document.getDocumentElement().normalize();
+        return document;
+    }
+
+    public static void getData(Document document) {
+        NodeList RowsTagList = document.getElementsByTagName("Row");
+
+        for (int i = 0; i < RowsTagList.getLength(); i++) {
+            Element row = (Element) RowsTagList.item(i);
+            NodeList ColsTagList = row.getElementsByTagName("Col");
+
+            for (int j = 0; j < ColsTagList.getLength(); j++) {
+                Element item = (Element) ColsTagList.item(j);
+                String id = item.getAttributes().getNamedItem("id").getNodeValue();
+
+                System.out.println(id + " " + item.getTextContent());
+            }
+        }
+    }
+}
```
XMLParser가 존재해야 하는 이유는 명확하다고 생각합니다. 하지만, 제가 위에서 말씀드린 예제를 보면 RestTemplate 쪽에서 이미 변환을 제공해주는 듯 해요.

youngxpepp reviewed (changes requested):
RestTemplate을 사용하는 것조차 놀라웠는데, MockTest까지 적용하시다니..
코드 정말 잘 봤습니다!

Do-ho committed:
refactor: 사용하지 않는 코드 삭제

Do-ho committed:
feat: RestTemplate Config 추가

- 계속적으로 연결을 시도하는 비용이 많이 들기에 Connection을 유지하는 RestTemplate를 유지하기로 결정

Do-ho committed:
feat: LectureRepository 생성

- Lecture에 대한 데이터를 기존 JpaRepository를 참조해 네이밍해 사용할 수 있도록 만듦

Do-ho committed:
feat: Lecture Builder 생성

- 기존 lecture들을 생성하기 위한 Builder 작성

Do-ho committed:
feat: LectureGetterService 작성

- XMLParser를 통해 얻어진 데이터를 Object로 바꾸기 위한 함수 작성
- Lecture를 생성하는 함수 작성

Do-ho committed:
refactor: Lecture 생성자 코드 수정

- lectureInfo를 통해 비즈니스 로직을 따로 관리하는 코드 작성

Do-ho committed:
feat: Lecture 비교 연산 및 lectureTime setter 구현

- 추후 lectureTime에 대한 비교 필요

Do-ho committed:
feat: LectureTime Builder 및 비즈니스 로직 작성

Do-ho committed:
feat: 통합 테스트 환경 수정

- 영속성을 위한 Transactinal 추가

Do-ho committed:
feat: LectureTimeRepository 생성

Do-ho committed:
feat: Test를 위한 상수 선언

Do-ho committed:
feat: LectureGetterService 로직 추가

- Lecture 생성자 선언 방식 변경에 따른 refactoring
- DB를 통해 가져온 list와 비교해 add, update, delete List 선정 후 Repository를 통해 변경
- 현재 LectureTime에 대한 비교가 어려워 해당 부분을 구현해야함

Do-ho committed:
feat: LectureGetterService Test 추가

Do-ho committed:
feat: Lecture 전체 목록 불러오는 Service 작성 및 테스트

Do-ho committed:
feat: cronScheduler 설정

- 5분 마다 lecture에 대한 업데이트 진행

Do-ho committed:
Merge branch 'develop' into feat/xmlParser

# Conflicts:
#	src/test/java/com/github/hanpyo/test/IntegrationTest.java

Do-ho committed:
Merge branch 'develop' into feat/xmlParser

Do-ho committed:
Merge branch 'feat/xmlParser' of https://github.com/koreatech/hanpyo-server into feat/xmlParser

Do-ho committed:
fix: Map 생성하는 부분 수정

- Java 9에서 사용되는 문법 수정

Do-ho committed:
fix: HashMap type 추가

Do-ho committed:
feat: LectureTime Entity 삭제 및 Time 로직 변경
